Code/GraphMol/MolProcessing/MolProcessing.cpp

```diff
@@ -0,0 +1,156 @@
+  std::map<unsigned int, std::unique_ptr<T>> accum;
+
+  auto workerfunc = [&](RWMol &mol, const std::string &,
+                        unsigned int recordId) {
+    auto item = func(mol);
+    {
+      std::lock_guard<std::mutex> lock(get_fp_mutex());
+      accum[recordId].reset(item);
+    }
+  };
+  suppl->setWriteCallback(workerfunc);
+  // loop over the supplier to make sure we read everything
+  while (!suppl->atEnd()) {
+    auto mol = suppl->next();
+  }
+  // convert the map to a vector and get the results in the input order
+  auto maxv = 0u;
+  for (const auto &pr : accum) {
+    maxv = std::max(maxv, pr.first);
+  }
+  results.resize(maxv);
+  for (auto &pr : accum) {
+    results[pr.first - 1] = std::move(pr.second);
+  }
+}
+#endif
+
+template <typename T>
+void worker(v2::FileParsers::MolSupplier *suppl,
+            std::function<T *(RWMol &)> func,
+            std::vector<std::unique_ptr<T>> &results) {
+  PRECONDITION(suppl, "no supplier");
+  // if we are using a multi-threaded supplier then we can register a write
+  // callback to do our processing multi-threaded too
+#ifdef RDK_BUILD_THREADSAFE_SSS
+  auto tsuppl =
+      dynamic_cast<v2::FileParsers::MultithreadedMolSupplier *>(suppl);
+  if (tsuppl) {
+    mtWorker(tsuppl, func, results);
+  } else {
+#else
+  {
+#endif
+    // otherwise we just loop through the molecules
+    while (!suppl->atEnd()) {
+      auto mol = suppl->next();
+      if (mol) {
+        auto fp = func(*mol);
+        results.emplace_back(fp);
+      } else if (!suppl->atEnd()) {
+        results.emplace_back(nullptr);
+      }
+    }
+  }
+}
+}  // namespace
+
+//! \brief Get fingerprints for all of the molecules in a file
+/*!
+   \param fileName the name of the file to read
+   \param options options controlling how the file is read, if not provided
+           four threads will be used whegn reading the file
+   \param generator the fingerprint generator to use, if not provided,
+           Morgan fingerprints with radius of 3 will be used.
+
+   \return an ExplicitBitVect,bitset pair, the first containing the
+           fingerprints and the second a bitset indicating which molecules were
+           successfully read
+*/
+template <typename OutputType>
+std::vector<std::unique_ptr<ExplicitBitVect>> getFingerprintsForMolsInFile(
+    const std::string &fileName,
+    const GeneralMolSupplier::SupplierOptions &options,
+    FingerprintGenerator<OutputType> *generator) {
+  auto suppl = details::getSupplier(fileName, options);
+
+  std::unique_ptr<FingerprintGenerator<OutputType>> morgan;
+  if (generator == nullptr) {
+    morgan.reset(MorganFingerprint::getMorganGenerator<OutputType>(3));
+    generator = morgan.get();
+  }
+  std::function<ExplicitBitVect *(RWMol &)> func = [&](RWMol &mol) {
+    return generator->getFingerprint(mol);
+  };
+  std::vector<std::unique_ptr<ExplicitBitVect>> results;
+  worker(suppl.get(), func, results);
+  return results;
+}
+
```

Comment:
It seems like this pattern will be used a lot.  Would a more functional interface be useful?  I.e.

```
template<Output> 
std::vector<Output> MolFileProcessing(std::string, func);
```

Something more like map and apply.

stupid example:

```
auto generator = std::make_unique<FingerprintGenerator<OutputType>>(3);
auto results = MolFileProcessing<std::unique_ptr<ExplicitBitVect>>(filename,
     [&](RWMol &mol) {
            return generator->getFingerprint(mol);
      };
);
```